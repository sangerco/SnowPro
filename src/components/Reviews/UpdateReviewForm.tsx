import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  fetchReview,
  updateReview,
  ReviewData,
} from "../../redux/slices/reviewSlice";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchAllTags } from "../../redux/slices/tagSlice";
import TagForm from "../Tags/TagForm";
import { useParams, useNavigate } from "react-router";
import {
  Container,
  Button,
  Dimmer,
  Loader,
  Form,
  Divider,
  Dropdown,
  DropdownItemProps,
  Header,
} from "semantic-ui-react";

const UpdateReviewForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { id } = useParams();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const username = auth.data?.username;
  const navigate = useNavigate();

  useEffect(() => {
    if (id) {
      dispatch(fetchReview(id));
    }
  }, [dispatch, id]);

  const reviews = useSelector((state: RootState) => state.reviews);
  const review = reviews.review;

  useEffect(() => {
    dispatch(fetchAllTags());
  }, [dispatch]);

  const tags = useSelector((state: RootState) => state.tags);

  let tagOptions:
    | { key: string; text: string; value: string }[]
    | DropdownItemProps[]
    | undefined;

  if (tags.tags) {
    tagOptions = tags.tags.map((tag) => ({
      key: tag.tagId,
      text: tag.tag,
      value: tag.tagId,
    }));
  } else {
    tagOptions = [];
  }

  const ratingOptions = [
    { key: 1, text: "1", value: 1 },
    { key: 2, text: "2", value: 2 },
    { key: 3, text: "3", value: 3 },
    { key: 4, text: "4", value: 4 },
    { key: 5, text: "5", value: 5 },
  ];

  let initialUpdateReviewState: ReviewData;

  if (review) {
    initialUpdateReviewState = {
      id: review.id,
      userId: review.userId,
      username: review.username,
      skiAreaSlug: review.skiAreaSlug,
      skiAreaName: review.skiAreaName,
      header: review.header,
      body: review.body,
      stars: review.stars,
      photos: review.photos || [],
      tags: review.tags || [],
      createdAt: review.createdAt,
    };
  } else {
    initialUpdateReviewState = {
      id: "",
      userId: userId || "",
      username: username || "",
      skiAreaSlug: "",
      skiAreaName: "",
      header: "",
      body: "",
      stars: 0,
      photos: [],
      tags: [],
      createdAt: new Date(),
    };
  }

  const [formData, setFormData] = useState(initialUpdateReviewState);
  const [photos, setPhotos] = useState(initialUpdateReviewState.photos);
  const [showCreateNewTagForm, setShowCreateNewTagForm] = useState(false);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(updateReview(formData));

    setFormData(initialUpdateReviewState);
  };

  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleTextAreaChange: React.ChangeEventHandler<HTMLTextAreaElement> = (
    e
  ) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handlePhotoChange = (index: number, value: string) => {
    if (formData.photos) {
      const updatedPhotoLinks = [...formData.photos];
      updatedPhotoLinks[index] = value;
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleAddPhotoInput = () => {
    if (formData.photos) {
      setPhotos([...formData.photos, ""]);
    }
  };

  const handleRemovePhotoInput = (index: number) => {
    if (formData.photos) {
      const updatedPhotoLinks = [...formData.photos];
      updatedPhotoLinks.splice(index, 1);
      setPhotos(updatedPhotoLinks);
    }
  };

  if (reviews.loading) {
    <Dimmer active>
      <Loader>Loading...</Loader>
    </Dimmer>;
  }

  if (reviews.error) {
    <Dimmer active>
      <Header as="h1">
        Error! Review cannot be retrieved! {reviews.error}
      </Header>
      <Button color="red" onClick={() => navigate(-1)}>
        Go Back.
      </Button>
    </Dimmer>;
  }

  if (review && userId !== review.userId) {
    return (
      <div>
        <Dimmer active>
          <Header as="h1">You are not authorized to edit this review.</Header>
          <Button color="red" onClick={() => navigate(-1)}>
            Go Back.
          </Button>
        </Dimmer>
      </div>
    );
  }

  if (review) {
    return (
      <Container fluid>
        <Form onSubmit={handleSubmit}>
          <Form.Field>
            <input
              placeholder={formData.body}
              name="header"
              value={formData.header}
              onChange={handleInputChange}
            />
          </Form.Field>
          <Form.TextArea
            placeholder={formData.body}
            name="body"
            value={formData.body}
            onChange={handleTextAreaChange}
          />
          {photos.map((photo, index) => (
            <Form.Field key={index}>
              <label>Photo Links</label>
              <input
                type="text"
                value={photo}
                onChange={(e) => handlePhotoChange(index, e.target.value)}
              />
              {index === photos.length - 1 && (
                <Button icon="plus" onClick={handleAddPhotoInput} />
              )}
              {index > 0 && (
                <Button
                  icon="minus"
                  onClick={() => handleRemovePhotoInput(index)}
                />
              )}
            </Form.Field>
          ))}
          <Divider />
          <label>Tags</label>
          <Dropdown
            clearable
            options={tagOptions}
            multiple
            fluid
            selection
            value={formData.tags}
          />
          <Button onClick={() => setShowCreateNewTagForm(true)} size="small">
            Create New Tag?
          </Button>
          {showCreateNewTagForm && <TagForm />}
          <Divider />
          <label>How would you rate your experience?</label>
          <Dropdown
            clearable
            options={ratingOptions}
            selection
            fluid
            value={formData.stars}
          />
          <Divider />
          <Button type="submit">Send it!</Button>
        </Form>
      </Container>
    );
  }

  return null;
};

export default UpdateReviewForm;
