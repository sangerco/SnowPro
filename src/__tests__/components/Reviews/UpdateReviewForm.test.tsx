/* eslint-disable testing-library/no-node-access */
import React from "react";
import UpdateReviewForm from "../../../components/Reviews/UpdateReviewForm";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Update Review Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewForm />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test updating review", () => {
  const reviewId = "review-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ reviewId });
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      review: {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created Date",
        replies: [],
      },
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .patch(`/api/reviews/review-1`, {
        skiAreaSlug: "ski-area-1",
      })
      .reply(201, {
        review: {
          id: "review-1",
          userId: "11",
          skiAreaSlug: "ski-area-1",
          header: "test review update",
          body: "test review body update",
          stars: 2,
          createdAt: "fake created date",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewForm />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const updateReviewForm = screen.getByTestId("update-review-form");
      expect(updateReviewForm).toBeInTheDocument();
    });

    await waitFor(() => {
      const reviewBody = screen.getAllByPlaceholderText("test review body");
      expect(reviewBody[0]).toBeInTheDocument();
    });

    await waitFor(() => {
      const submitButton = screen.getByTestId("review-button");
      expect(submitButton).toBeInTheDocument();
    });
  });
  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .patch(`/api/reviews/review-1`, {
        skiAreaSlug: "ski-area-1",
      })
      .reply(201, {
        review: {
          id: "review-1",
          userId: "11",
          skiAreaSlug: "ski-area-1",
          header: "test review update",
          body: "test review body update",
          stars: 2,
          createdAt: "fake created date",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewForm />
        </BrowserRouter>
      </Provider>
    );

    expect(nock.pendingMocks().length).toBe(3);

    await waitFor(() => {
      const submitButton = screen.getByTestId("review-button");
      expect(submitButton).toBeInTheDocument();
    });
  });
});
