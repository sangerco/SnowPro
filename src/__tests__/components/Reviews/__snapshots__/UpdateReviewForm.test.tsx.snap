// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Nav Bar should match snapshot 1`] = `<DocumentFragment />`;

exports[`test Update Review Form should match snapshot 1`] = `<DocumentFragment />`;
